workers/unity/Assets/Gdk/Core/Components/Option.cs

```diff
@@ -0,0 +1,120 @@
+using System;
+
+namespace Improbable.Gdk.Core
+{
+    public struct Option<T> : IEquatable<Option<T>>
+    {
+        private bool hasValue;
+        private T value;
+        
+        public Option(T newValue)
+        {
+            hasValue = false;
+            value = default(T);
+            Value = newValue;
+        }
+
+        public BlittableBool HasValue => hasValue;
+
+        public T Value
+        {
+            get
+            {
+                if (!hasValue)
+                {
+                    throw new InvalidOperationException("Called Value on empty Option.");
+                }
+
+                return value;
+            }
+            set
+            {
+                if (value == null)
```

Comment:
This boxes I think? (value: `T` -> `object`) since `T` can be a value type

Reply:
You might want to make this something other than a type T. Possibly an internal type than can be implicitly cast to T but can also be set with null (basically a nullable T wrapper). It would create the oddity that you can have an option ref with a value but that value being null. But that might be fine.

We could also try branching off `typeof(T).isValueType` and having things deal with the two different cases there. Either way it might be better if `Value` is garunteed to be a type that cab be assigned null.

Reply:
I'd mildly prefer not having a setter here at all though. Users shouldn't ever be setting value directly. It's just an accessor that can throw if there is no value present.

Reply:
Discussed offline: Made Options immutable and made empty options distinct from options with payload null.